### PR TITLE
NMS-14667: Official docs readiness for Cortex TSS plugin release

### PR DESCRIPTION
**Pull request by Bonrob2**
### All Contributors

* [ ] Have you read our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [ ] Have you (electronically) signed the [OpenNMS Contributor Agreement](https://cla-assistant.io/OpenNMS/opennms)?

### Contribution Checklist

* Please [make an issue in the OpenNMS issue tracker](https://issues.opennms.org) if there isn't one already.<br />Once there is an issue, please:
  1. update the title of this PR to be in the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
  4. once you've created this PR, please link to it in a comment in the JIRA issue
  Don't worry if this sounds like a lot, we can help you get things set up properly.
* **If this code is likely to affect the UI, did you name your branch with `-smoke` in it to trigger smoke tests?**
* If this is a new or updated feature, is there documentation for the new behavior?
* If this is new code, are there unit and/or integration tests?
* If this PR targets a `foundation-*` branch, does it try to avoid changing files in `$OPENNMS_HOME/etc/`?

### What's Next?

A PR should be assigned at least 2 reviewers.  If you know that someone would be a good person to review your code, feel free to add them.

If you need help making additions or changes to the documentation related to your changes, please let us know.

In any case, if anything is unclear or you want help getting your PR ready for merge, please don't hesitate to say something in the comments here,
or in [the #opennms-development chat channel](https://chat.opennms.com/opennms/channels/opennms-development).

Once reviewer(s) accept the PR and the branch passes continuous integration, the PR is eligible for merge.

At that time, if you have commit access (are an OpenNMS Group employee or a member of the OGP) you are welcome to merge the PR when you're ready.
Otherwise, a reviewer can merge it for you.

Thanks for taking time to contribute!

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-14667

### CODE REVIEW

**Comment by Bonrob2** [comment]
Note, @jeffgdotorg that the image is a placeholder for now. Also, I am not sure what the Cortex tips section is supposed to help with. The links don't work, and there is no explanation in the readme.

**Review by indigo423 — COMMENTED** [review]
Do we have the yED source file for the cortex.png somewhere? If someone needs to change it later, it would require completely rebuilding it?

**Comment by Bonrob2** [comment]
Yes, we will have it in yED. Right now, this is a placeholder. I didn't want to include the current original version as it hasn't been approved yet.
________________________________

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 1]
```diff
@@ -0,0 +1,91 @@
+
+= Cortex Time Series Plugin
+
+The Time Series Storage Plugin for Cortex helps {page-component-title} to persist aggregated flow data to Cortex.
```
Comment (jeffgdotorg) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:4]: Let's change "helps" here to "enables".

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+The plugin uses the xref:deployment:time-series-storage/timeseries/ts-integration-layer.adoc[time series storage integration layer] to convert metrics to a Prometheus model and delegate writes and reads to https://cortexmetrics.io/[Cortex].
+The plugin also makes it possible for https://docs.opennms.com/helm/latest/index.html[OpenNMS Helm] to visualize flow graphs from the Cortex instance.
```
Comment (jeffgdotorg) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:6]: Let's be more specific about which part of Helm. It's the Helm Performance Data Source Plugin.

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+== Deployment
+
+To use the Time Series Storage Plugin for Cortex, do the following:
+
+.Start Cortex
+See https://cortexmetrics.io/docs/getting-started/.
+
+You can also download a https://github.com/opennms-forge/stack-play/tree/master/standalone-cortex-minimal[minimal, standalone version of Cortex] and start Cortex with `docker-compose up`.
+
+.Build and install the plugin into your local Maven repository
```
Comment (jeffgdotorg) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:30]: We should be able to eliminate this developer-centric step, replacing it with instructions for downloading the pre-built plugin from GitHub or another place. Details will become available closer to the September release day.

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+property-set writeTimeoutInMs 1000
+property-set readTimeoutInMs 1000
+property-set metricCacheSize 1000
+property-set externalTagsCacheSize 1000
+property-set bulkheadMaxWaitDurationInMs 9223372036854775807
+
+config:update
+----
+
+.Update automatically
```
Comment (jeffgdotorg) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:76]: This step is specific to developer environments, and should also go away once we're building and releasing the plugin.

**Review by jeffgdotorg — CHANGES_REQUESTED** [review]
Only my first two comments require action at this time. The rest are advisory.

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc (hunk 2 above)
Comment (dino2gnt) [reply docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:6]: >visualize flow graphs from the Cortex instance.

>It's the Helm Performance Data Source

Is it?

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc (hunk 3 above)
Comment (Bonrob2) [reply docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:30]: Do we have a general idea of how this will work so that we can rough it in?

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc (hunk 2 above)
Comment (Bonrob2) [reply docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:6]: I'll update the Helm sentence to be less project based, and more user based. Users can visualize flow graphs from Cortex in Helm through the performance data source. Something like that.

Comment (Bonrob2) [reply docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:6]: @jeffgdotorg is there additional configuration required in Helm to make things work with Cortex? If so, we need to document it.

**Comment by j-white** [comment]
This plugin is strictly used for time series - it has nothing to do with flows.

(Nephron on the other hand can be configured to persist to Cortex, and the results can then be visualized using the Prometheus datasource. This plugin, or even Helm aren't at play in this scenario though.)

**Comment by Bonrob2** [comment]
> This plugin is strictly used for time series - it has nothing to do with flows.
> 
> (Nephron on the other hand can be configured to persist to Cortex, and the results can then be visualized using the Prometheus datasource. This plugin, or even Helm aren't at play in this scenario though.)

Thanks for the clarification, Jesse. I included the Helm bit because one of the Jiras associated with this feature has the following acceptance criteria: "the OpenNMS Helm application can be used to visualize performance data persisted to the Cortex DB".

**Comment by Bonrob2** [comment]
> This plugin is strictly used for time series - it has nothing to do with flows.
> 
> (Nephron on the other hand can be configured to persist to Cortex, and the results can then be visualized using the Prometheus datasource. This plugin, or even Helm aren't at play in this scenario though.)

Will update.

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+* Cortex version 1.10.0+
+
+NOTE: We plan to keep the plugin compatible with https://grafana.com/oss/mimir/[Mimir], Grafana's fork of Cortex, as long as Mimir remains compatible with Cortex.
+The minimum supported Mimir version is 2.0.0.
+
+== Deployment
+
+To use the Time Series Storage Plugin for Cortex, do the following:
+
+. Start Cortex
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:25]:
```suggestion
. Start Cortex.
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
++
+See https://cortexmetrics.io/docs/getting-started/.
++
+You can also download a https://github.com/opennms-forge/stack-play/tree/master/standalone-cortex-minimal[minimal, standalone version of Cortex] and start Cortex with `docker-compose up`.
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:29]:
```suggestion
You can also download a https://github.com/opennms-forge/stack-play/tree/master/standalone-cortex-minimal[minimal, standalone version of Cortex] and start Cortex with the `docker-compose up` command.
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+. Download the plugin
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:31]:
```suggestion
. Download the plugin.
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
++
+TBD
+
+. Enable and configure the time series storage
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:35]:
```suggestion
. Enable and configure the time series storage.
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+----
+echo "org.opennms.timeseries.strategy=integration
+org.opennms.timeseries.tin.metatags.tag.node=${node:label}
+org.opennms.timeseries.tin.metatags.tag.location=${node:location}
+org.opennms.timeseries.tin.metatags.tag.geohash=${node:geohash}
+org.opennms.timeseries.tin.metatags.tag.ifDescr=${interface:if-description}
+org.opennms.timeseries.tin.metatags.tag.label=${resource:label}" >> ${OPENNMS_HOME}/etc/opennms.properties.d/cortex.properties
+----
+
+. Install the plugin from the OpenNMS Karaf shell
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:47]:
```suggestion
. Install the plugin from the OpenNMS Karaf shell.
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
++
+[source, console]
+----
+feature:repo-add mvn:org.opennms.plugins.timeseries/cortex-karaf-features/1.0.0-SNAPSHOT/xml
+feature:install opennms-plugins-cortex-tss
+----
+
+. Configure
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:55]:
```suggestion
. Configure the plugin.
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
++
+Note that you can omit this step if you use the default values.
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:57]:
```suggestion
Note that you can omit this step if you use the default settings.
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+== Cortex tips
+
+=== View the ring
+
+http://localhost:9009/ring
+
+=== View internal metrics
+
+http://localhost:9009/metrics
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:83]: Should these sections be expanded to include more detail?

File: docs/modules/operation/pages/database-reports/database.adoc [hunk 13]
```diff
@@ -176,7 +176,7 @@ The official Jaspersoft Studio documentation and wiki describe how to do this.
 To use the Measurements API you must add the Measurements Datasource library to the build path of JasperStudio.
 To do so, right click in the Project Explorer and select *Configure Buildpath*.
```
Comment (emarsh1995) [inline docs/modules/operation/pages/database-reports/database.adoc:177]:
```suggestion
To do so, right-click in the Project Explorer and select *Configure Buildpath*.
```

**Review by emarsh1995 — CHANGES_REQUESTED** [review]

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc (hunk 12 above)
Comment (Bonrob2) [reply docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:83]: Yes. We will add more information closer to the release. I'll update, though, to make that clear.

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 14]
```diff
@@ -0,0 +1,83 @@
+
+= Cortex Time Series Plugin
+
+The Time Series Storage Plugin for Cortex enables {page-component-title} to persist aggregated flow data to Cortex.
```
Comment (jeffgdotorg) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:4]: Please change "aggregated flow data" to "time-series performance and latency data".

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+
+== Deployment
+
+To use the Time Series Storage Plugin for Cortex, do the following:
+
+. Start Cortex.
++
+See https://cortexmetrics.io/docs/getting-started/.
++
+You can also download a https://github.com/opennms-forge/stack-play/tree/master/standalone-cortex-minimal[minimal, standalone version of Cortex] and start Cortex with the `docker-compose up` command.
```
Comment (jeffgdotorg) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:29]: Let's rewrite to:
If your aim is to get up to speed quickly in a test environment, you can use Docker Compose to deploy a https://github.com/opennms-forge/stack-play/tree/master/standalone-cortex-minimal[minimal, standalone Cortex instance].

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+property-set bulkheadMaxWaitDurationInMs 9223372036854775807
+
+config:update
+----
+
+== Cortex tips
+
+=== View the ring
+
+http://localhost:9009/ring
```
Comment (jeffgdotorg) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:79]: Let's change to `http://cortex-service-ip-addr:9009/ring`

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+
+=== View internal metrics
+
+http://localhost:9009/metrics
```
Comment (jeffgdotorg) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:83]: Let's change to `http://cortex-service-ip-addr:9009/metrics`

**Review by jeffgdotorg — CHANGES_REQUESTED** [review]

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc (hunk 14 above)
Comment (Bonrob2) [reply docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:4]:
```suggestion
The Time Series Storage Plugin for Cortex enables {page-component-title} to persist time-series performance and latency data to Cortex.
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc (hunk 15 above)
Comment (Bonrob2) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:29]:
```suggestion
If your aim is to get up to speed quickly in a test environment, you can use Docker Compose to deploy a https://github.com/opennms-forge/stack-play/tree/master/standalone-cortex-minimal[minimal, standalone version of Cortex].
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc (hunk 16 above)
Comment (Bonrob2) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:79]:
```suggestion
http://cortex-service-ip-addr:9009/ring
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc (hunk 17 above)
Comment (Bonrob2) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:83]:
```suggestion
http://cortex-service-ip-addr:9009/metrics
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+org.opennms.timeseries.tin.metatags.tag.geohash=${node:geohash}
+org.opennms.timeseries.tin.metatags.tag.ifDescr=${interface:if-description}
+org.opennms.timeseries.tin.metatags.tag.label=${resource:label}" >> ${OPENNMS_HOME}/etc/opennms.properties.d/cortex.properties
+----
+
+. Install the plugin from the OpenNMS Karaf shell.
++
+[source, console]
+----
+feature:repo-add mvn:org.opennms.plugins.timeseries/cortex-karaf-features/1.0.0-SNAPSHOT/xml
```
Comment (j-white) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:49]: We can remove this step, if we assume that the users have installed the `.kar` file as noted above.

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+feature:install opennms-plugins-cortex-tss
+----
+
```
Comment (j-white) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:52]: There's an additional step to take to make sure that the plugin remains installed/started after reboot. This would look like:
```
echo 'opennms-plugins-cortex-tss wait-for-kar=opennms-cortex-plugin' > "$OPENNMS_HOME/etc/featuresBoot.d/cortex.boot"
```
(Needs to be validated)

Comment (Bonrob2) [reply docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:52]: @jeffgdotorg  or @j-white is there someone who can validate this? We need to merge this today.

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+== Deployment
+
+To use the Cortex Time Series Plugin, do the following:
+
+. Start Cortex.
++
+See https://cortexmetrics.io/docs/getting-started/.
++
+If your aim is to get up to speed quickly in a test environment, you can use Docker Compose to deploy a https://github.com/opennms-forge/stack-play/tree/master/standalone-cortex-minimal[minimal, standalone version of Cortex].
```
Comment (jeffgdotorg) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:29]: Let's change link text to "minimal, standalone Cortex cluster".

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+
+. Download the plugin's `.kar` file from GitHub (https://github.com/OpenNMS/opennms-cortex-tss-plugin/releases).
+. Copy the file into the `opennms/deploy` directory of your running instance.
```
Comment (jeffgdotorg) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:32]: Please change `opennms/deploy` to `OPENNMS_HOME/deploy`, or whatever convention we use elsewhere to denote the directory that's `/opt/opennms/` on RPM systems or `/usr/share/opennms` on Debian-based ones.

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+. Enable and configure the time series storage.
```
Comment (jeffgdotorg) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:33]: Please change to "Enable and configure time series storage plugins in {page-component-title}"

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+----
+echo "org.opennms.timeseries.strategy=integration
+org.opennms.timeseries.tin.metatags.tag.node=${node:label}
+org.opennms.timeseries.tin.metatags.tag.location=${node:location}
+org.opennms.timeseries.tin.metatags.tag.geohash=${node:geohash}
+org.opennms.timeseries.tin.metatags.tag.ifDescr=${interface:if-description}
+org.opennms.timeseries.tin.metatags.tag.label=${resource:label}" >> ${OPENNMS_HOME}/etc/opennms.properties.d/cortex.properties
+----
+
+. Configure the plugin.
```
Comment (jeffgdotorg) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:45]: Please change to "Configure the Cortex time series storage plugin in {page-component-title}"

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
++
+Note that you can omit this step if you use the default settings.
```
Comment (jeffgdotorg) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:47]: Please add a note here: "Substitute the correct host and port for your Cortex cluster in each place where `cortex-service-host:9009` appears below."

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
++
+[source, console]
+----
+config:edit org.opennms.plugins.tss.cortex
+
+property-set writeUrl http://localhost:9009/api/prom/push
```
Comment (jeffgdotorg) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:53]: In each line of this source block, please change `localhost` to `cortex-service-host`, so that the URL base is `http://cortex-service-host:9009` throughout.

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+property-set maxConcurrentHttpConnections 100
+property-set writeTimeoutInMs 1000
+property-set readTimeoutInMs 1000
+property-set metricCacheSize 1000
+property-set externalTagsCacheSize 1000
+property-set bulkheadMaxWaitDurationInMs 9223372036854775807
+
+config:update
+----
+. Configure the plugin to remain installed after reboot:
```
Comment (jeffgdotorg) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:64]: Please change "after reboot" to "after a restart of {page-component-title}".

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 27]
```diff
@@ -0,0 +1,79 @@
...
+
+[source,console]
+----
+echo 'opennms-plugins-cortex-tss wait-for-kar=opennms-cortex-plugin' > "$OPENNMS_HOME/etc/featuresBoot.d/cortex.boot"
+----
+
+== Cortex tips
```
Comment (jeffgdotorg) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:71]: For consistency with my changes above, please also change `cortex-service-ip-addr` below to `cortex-service-host`.

File: docs/modules/operation/pages/database-reports/database.adoc [hunk 28]
```diff
@@ -174,9 +174,9 @@ The official Jaspersoft Studio documentation and wiki describe how to do this.
 === Use Measurements Datasource and Helpers
 
 To use the Measurements API you must add the Measurements Datasource library to the build path of JasperStudio.
-To do so, right click in the Project Explorer and select *Configure Buildpath*.
```
Comment (jeffgdotorg) [inline docs/modules/operation/pages/database-reports/database.adoc:177]: Does this file's changes belong in this PR?

File: features/themes/onms-default-theme/bin/pom.xml [hunk 29]
```diff
@@ -0,0 +1,107 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jeffgdotorg) [inline features/themes/onms-default-theme/bin/pom.xml:1]: I'm pretty sure this file does not belong in this PR.

File: features/themes/onms-default-theme/bin/src/license/THIRD-PARTY.properties [hunk 30]
```diff
@@ -0,0 +1,2 @@
+dom4j--dom4j--1.6.1=DOM4J
```
Comment (jeffgdotorg) [inline features/themes/onms-default-theme/bin/src/license/THIRD-PARTY.properties:1]: I'm pretty sure this file does not belong in this PR.

File: features/themes/onms-default-theme/bin/src/main/resources/VAADIN/themes/opennms/styles.css [hunk 31]
```diff
@@ -0,0 +1 @@
+@import "./assets/vaadin-theme.css?v=${buildNumber}";
```
Comment (jeffgdotorg) [inline features/themes/onms-default-theme/bin/src/main/resources/VAADIN/themes/opennms/styles.css:1]: I'm pretty sure this file does not belong in this PR.

**Review by jeffgdotorg — CHANGES_REQUESTED** [review]

File: docs/modules/operation/pages/database-reports/database.adoc (hunk 28 above)
Comment (Bonrob2) [reply docs/modules/operation/pages/database-reports/database.adoc:177]: It isn't incorrect, so we can keep it.

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 32]
```diff
@@ -0,0 +1,80 @@
+
+= Cortex Time Series Plugin
+
+The Cortex Time Series Plugin enables {page-component-title} to persist time-series performance and latency data to Cortex.
+The plugin uses the xref:deployment:time-series-storage/timeseries/ts-integration-layer.adoc[time series storage integration layer] to convert metrics to a Prometheus model and delegate writes and reads to https://cortexmetrics.io/[Cortex].
+The plugin also makes it possible for users to visualize graphs from the Cortex instance in https://docs.opennms.com/helm/latest/index.html[OpenNMS Helm], through the OpenNMS performance data source or the https://grafana.com/grafana/plugins/prometheus/[Prometheus Data Source].
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:6]:
```suggestion
The plugin also lets users visualize graphs from the Cortex instance in https://docs.opennms.com/helm/latest/index.html[OpenNMS Helm] through the OpenNMS performance data source or the https://grafana.com/grafana/plugins/prometheus/[Prometheus data source].
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 33]
```diff
@@ -0,0 +1,80 @@
+
+= Cortex Time Series Plugin
+
+The Cortex Time Series Plugin enables {page-component-title} to persist time-series performance and latency data to Cortex.
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:4]:
```suggestion
The Cortex Time Series plugin enables {page-component-title} to persist time series performance and latency data to Cortex.
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 34]
```diff
@@ -0,0 +1,80 @@
...
+The plugin uses the xref:deployment:time-series-storage/timeseries/ts-integration-layer.adoc[time series storage integration layer] to convert metrics to a Prometheus model and delegate writes and reads to https://cortexmetrics.io/[Cortex].
+The plugin also makes it possible for users to visualize graphs from the Cortex instance in https://docs.opennms.com/helm/latest/index.html[OpenNMS Helm], through the OpenNMS performance data source or the https://grafana.com/grafana/plugins/prometheus/[Prometheus Data Source].
+
+.Cortex Time Series Plugin
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:8]:
```suggestion
.Cortex Time Series plugin
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 35]
```diff
@@ -0,0 +1,80 @@
...
+
+* Horizon 30.0.0+ (Meridian availability with the 2023 release)
+* Cortex version 1.10.0+
+
+NOTE: We plan to keep the plugin compatible with https://grafana.com/oss/mimir/[Mimir], Grafana's fork of Cortex, as long as Mimir remains compatible with Cortex.
+The minimum supported Mimir version is 2.0.0.
+
+== Deployment
+
+To use the Cortex Time Series Plugin, do the following:
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:23]:
```suggestion
To use the Cortex Time Series plugin, do the following:
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 36]
```diff
@@ -0,0 +1,80 @@
...
+
+. Start Cortex.
++
+See https://cortexmetrics.io/docs/getting-started/.
++
+If your aim is to get up to speed quickly in a test environment, you can use Docker Compose to deploy a https://github.com/opennms-forge/stack-play/tree/master/standalone-cortex-minimal[minimal, standalone Cortex cluster].
+
+. Download the plugin's `.kar` file from GitHub (https://github.com/OpenNMS/opennms-cortex-tss-plugin/releases).
+. Copy the file into the `$\{OPENNMS_HOME}/deploy` directory of your running instance.
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:32]:
```suggestion
. Copy the file into `$\{OPENNMS_HOME}/deploy`.
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 37]
```diff
@@ -0,0 +1,80 @@
...
+. Enable and configure time series storage plugins in {page-component-title}.
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:33]:
```suggestion
. Enable and configure time series storage plugins in {page-component-title}:
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 38]
```diff
@@ -0,0 +1,80 @@
...
+----
+echo "org.opennms.timeseries.strategy=integration
+org.opennms.timeseries.tin.metatags.tag.node=${node:label}
+org.opennms.timeseries.tin.metatags.tag.location=${node:location}
+org.opennms.timeseries.tin.metatags.tag.geohash=${node:geohash}
+org.opennms.timeseries.tin.metatags.tag.ifDescr=${interface:if-description}
+org.opennms.timeseries.tin.metatags.tag.label=${resource:label}" >> ${OPENNMS_HOME}/etc/opennms.properties.d/cortex.properties
+----
+
+. Configure the Cortex time series storage plugin in {page-component-title}.
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:45]:
```suggestion
. Configure the Cortex Time Series storage plugin in {page-component-title}.
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 39]
```diff
@@ -0,0 +1,80 @@
...
+property-set maxConcurrentHttpConnections 100
+property-set writeTimeoutInMs 1000
+property-set readTimeoutInMs 1000
+property-set metricCacheSize 1000
+property-set externalTagsCacheSize 1000
+property-set bulkheadMaxWaitDurationInMs 9223372036854775807
+
+config:update
+----
+. Configure the plugin to remain installed after a restart of {page-component-title}:
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:65]:
```suggestion

. Configure the plugin to remain installed after a restart of {page-component-title}:
```

File: docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc [hunk 40]
```diff
@@ -0,0 +1,80 @@
...
+
```
Comment (emarsh1995) [inline docs/modules/deployment/pages/time-series-storage/timeseries/cortex.adoc:66]:
```suggestion
+
```

**Review by emarsh1995 — CHANGES_REQUESTED** [review]
Just a few minor nits. :)

**Review by jeffgdotorg — APPROVED** [review]
Ship it! :shipit:

**Review by emarsh1995 — APPROVED** [review]
LGTM! Thanks! :D
